gh-99108: Refactor _sha256 & _sha512 into _sha2.

gpshead opened the pull request:
This merges their code. They're backed by the same single HACL* static library, having them be a single module simplifies maintenance.

This should unbreak the wasm enscripten builds that currently fail due to linking in --whole-archive mode and the HACL* library appearing twice.

Long unnoticed error fixed: _sha512.SHA384Type was doubly assigned and was actually SHA512Type. Nobody depends on those internal names.

This is a followup to PRs #101707 and #101917.

<!-- gh-issue-number: gh-99108 -->
* Issue: gh-99108
<!-- /gh-issue-number -->

gpshead committed:
gh-99108: Refactor _sha256 & _sha512 into _sha2.

This merges their code. They're backed by the same single HACL* star static
library, having them be a single module simplifies maintenance.

This should unbreak the wasm enscripten builds that currently fail due to
linking in --whole-archive mode and the HACL* library appearing twice.

Long unnoticed error fixed: `_sha512.SHA384Type` was doubly assigned and was
actually SHA512Type. Nobody depends on those internal names.

gpshead committed:
update stdlib module names list.

gpshead committed:
remove the added subTest calls (no real value).

erlend-aasland commented on configure.ac:
```diff
@@ -7197,11 +7196,8 @@ dnl By default we always compile these even when OpenSSL is available
-  [-I\$(srcdir)/Modules/_hacl/include -I\$(srcdir)/Modules/_hacl/internal -D_BSD_SOURCE -D_DEFAULT_SOURCE])
-PY_STDLIB_MOD([_sha512],
-  [test "$with_builtin_sha512" = yes], [],
+PY_STDLIB_MOD([_sha2],
+  [test "$with_builtin_sha2" = yes], [],
```
Nit: these lines can now be joined :)

gpshead replied:
other things seem to keep them on multiple lines wen adding compiler or linker flags so I'll leave it this way.  m4 "style" is purely "be consistent with what I see nearby" for me. ;)

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+typedef struct {
+  PyObject_HEAD
+  int digestsize;
+  Hacl_Streaming_SHA2_state_sha2_256 *state;
```
Nit: 4 space indent.

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+static inline sha2_state*
+sha2_get_state(PyObject *module)
+{
+    void *state = PyModule_GetState(module);
```
```suggestion
    void *state = _PyModule_GetState(module);
```

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+{
+    SHA256object *sha = (SHA256object *)PyObject_GC_New(SHA256object,
+                                                  state->sha224_type);
+    PyObject_GC_Track(sha);
```
IIRC, `PyObject_GC_Track` will segfault if given a `NULL` pointer. We should do some error checking post `PyObject_GC_New`.

Ditto for the other similar functions.

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+static void
+SHA256_dealloc(SHA256object *ptr)
+{
+    Hacl_Streaming_SHA2_free_256(ptr->state);
```
Untrack first? Perhaps it does not matter.

gpshead replied:
I don't think it matters here, the GIL is held. (long existing code)

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+        GET_BUFFER_VIEW_OR_ERROUT(string, &buf);
+    }
+
+    sha2_state *state = PyModule_GetState(module);
```
```suggestion
    sha2_state *state = sha2_get_state(module);
```

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+        GET_BUFFER_VIEW_OR_ERROUT(string, &buf);
+    }
+
+    sha2_state *state = PyModule_GetState(module);
```
```suggestion
    sha2_state *state = sha2_get_state(module);
```

erlend-aasland commented on Modules/sha2module.c:
```diff
@@ -0,0 +1,789 @@
+
+    Py_INCREF((PyObject *)state->sha224_type);
+    if (PyModule_AddObject(module, "SHA224Type", (PyObject *)state->sha224_type) < 0) {
+        Py_DECREF((PyObject *)state->sha224_type);
```
If you use `PyModule_AddType`, you can get rid of the unneeded refcount dance here. Ditto for the other types below.

gpshead replied:
this required changing the type tp_name values to match, but that's probably a good idea anyways.

msprotz commented on Misc/NEWS.d/next/Library/2023-02-15-01-54-06.gh-issue-99108.rjTSic.rst:
```diff
@@ -0,0 +1,3 @@
+The built-in extension modules for :mod:`hashlib` SHA2 algorithms, used when
+OpenSSL does provide them, now live in a single internal ``_sha2`` module
```
does not?

gpshead committed:
cleanup per review comments.

gpshead committed:
rename the _sha2 types and simplify module init to use PyModule_AddType.

gpshead committed:
Explicitly test that direct type constructors fail.

gpshead committed:
missing param

(this is how useful github vscode editor w/o ability to build and test before commit is...)

gpshead committed:
remove the unconstructable test (not a valid constraint for everything)

gpshead committed:
Merge branch 'main' into build/sha2

gpshead committed:
fix compiler warnings.

gpshead committed:
Merge branch 'build/sha2' of github.com:gpshead/cpython into build/sha2

gpshead committed:
Merge remote-tracking branch 'upstream/main' into build/sha2

gpshead committed:
rename LIBHACL_ make vars to LIBHACL_SHA2_
